### PR TITLE
SLING-8271 implement AbstractStoppableDepthFirstVisitor

### PR DESCRIPTION
**Pull request by kwin**

### CODE REVIEW

File: src/main/java/org/apache/sling/api/resource/AbstractResourceVisitor.java [hunk 1]
```diff
@@ -23,13 +23,13 @@
...
- * This visitor will traverse the given resource and all its children in a breadth-first approach
+ * This visitor will traverse the given resource and all its children in a depth-first approach
  * and call the {@link AbstractResourceVisitor#visit(Resource)} method for each visited resource.
  * It decouples the actual traversal code from application code. 
  * 
  * Concrete subclasses must implement the {@link AbstractResourceVisitor#visit(Resource)} method.
  *
- * @see <a href="https://en.wikipedia.org/wiki/Breadth-first_search">Breadth-First-Search</a>
+ * @see <a href="https://en.wikipedia.org/wiki/Depth-first_search">Depth-First-Search</a>
  * @since 2.2 (Sling API Bundle 2.2.0)
```
Comment (rombert) [inline src/main/java/org/apache/sling/api/resource/AbstractResourceVisitor.java]: I would suggest making this part of a different change, it's important enough to stand out.

File: src/main/java/org/apache/sling/api/resource/AbstractStoppableDepthFirstResourceVisitor.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    /**
+     * Defines how the traversal should be continued.
+     * <ul>
+     * <li>{@link #NORMAL} for regular continuation</li>
+     * <li>{@link #SKIP_SUBTREE} to skip the children of the current resource and continue with its siblings</li>
+     * <li>{@link #STOP} to stop the whole traversal (i.e. to no visit any other resource afterwards)</li>
+     */
+    enum TraversalContinuation {
+        NORMAL,
```
Comment (rombert) [inline src/main/java/org/apache/sling/api/resource/AbstractStoppableDepthFirstResourceVisitor.java]: I would like to propose a set of alternate names. For your consideration only, I don't necessarily view them as "better"

- CONTINUE
- SKIP_CHILDREN
- SKIP_ALL

They sound a bit more natural to me, however it's naming so things are very subjective here.

**Comment by kwin** [comment]
As outlined in https://issues.apache.org/jira/browse/SLING-8271 the ResourceStream ( https://github.com/apache/sling-org-apache-sling-resource-filter/blob/master/src/main/java/org/apache/sling/resource/filter/ResourceStream.java) and ResourceFilter API is the better approach these days. Therefore I am closing this PR here unmerged.

File: src/main/java/org/apache/sling/api/resource/AbstractResourceVisitor.java (hunk 1 above)
Comment (kwin) [reply src/main/java/org/apache/sling/api/resource/AbstractResourceVisitor.java]: Will commit this separately

**Pull request closed without merging**